Moved IBOLocalizer to Libmints from FISAPT

andyj10224 opened the pull request:
## Description
Addresses issue #2379. IBOLocalizer is a MO localizer that exists in the FISAPT branch. Moving it to libmints cleans up the code.

## Checklist
- [ ] All fsapt tests run
- [ ] FSAPT tests passing
- [ ] FSAPT-D tests passing

## Status
- [ ] Ready for review
- [ ] Ready for merge

andyj10224 commented:
This PR addresses #2379 

All fsapt tests pass after refactoring (except fsapt-d) Any insights? This is the error message:
`!----------------------------------------------------------------------------------!
!                                                                                  !
!  Unsuccessful run. Check input, particularly geometry in [a0]. Model:            !
!     Model(method='hf-d3', basis='(auto)')                                        !
!                                                                                  !
!----------------------------------------------------------------------------------!`

hokru commented:
Could you post the whole test output please?

And do you have the `dftd3` binary available at run time?

andyj10224 commented:
> Could you post the whole test output please?
> 
> And do you have the `dftd3` binary available at run time?

Yep, it was a dftd3 issue. Rerunning tests now.

andyj10224 commented:
All tests passing. PR ready for review.

susilehtola commented:
IBO could be combined with Pipek-Mezey; the only difference is the way the orbital overlap is computed.

JonathonMisiewicz commented on psi4/src/psi4/fisapt/fisapt.cc:
```diff
@@ -118,8 +118,8 @@ void FISAPT::localize() {
     ranges.push_back(vectors_["eps_focc"]->dimpi()[0]);
     ranges.push_back(vectors_["eps_occ"]->dimpi()[0]);
 
-    std::shared_ptr<fisapt::IBOLocalizer2> local =
```
While this is tangential to the purpose of the PR, please do the trivial cleanup of using `auto` in all three code blocks where the variable is defined.

JonathonMisiewicz commented on psi4/src/psi4/fisapt/fisapt.cc:
```diff
@@ -3106,8 +3106,8 @@ void FISAPT::flocalize() {
             new Matrix("Focc", vectors_["eps_occ0A"]->dimpi()[0], vectors_["eps_occ0A"]->dimpi()[0]));
```
While you're here, please change to use `make_shared`. Do the same on 3174.

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+    common_init();
+}
+IBOLocalizer::~IBOLocalizer() {}
+void IBOLocalizer::common_init() {
```
`common_init` is not marked `override`. Is it meant to override the base class constructor or not?

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+
+    auto local = std::make_shared<IBOLocalizer>(primary, minao, C);
+
+    local->set_print(options.get_int("PRINT"));
+    local->set_debug(options.get_int("DEBUG"));
+    local->set_bench(options.get_int("BENCH"));
+    local->set_convergence(options.get_double("LOCAL_CONVERGENCE"));
+    local->set_maxiter(options.get_int("LOCAL_MAXITER"));
+    local->set_use_ghosts(options.get_bool("LOCAL_USE_GHOSTS"));
+    local->set_condition(options.get_double("LOCAL_IBO_CONDITION"));
+    local->set_power(options.get_double("LOCAL_IBO_POWER"));
+    local->set_use_stars(options.get_bool("LOCAL_IBO_USE_STARS"));
+    local->set_stars_completeness(options.get_double("LOCAL_IBO_STARS_COMPLETENESS"));
+
+    std::vector<int> stars;
+    for (int ind = 0; ind < options["LOCAL_IBO_STARS"].size(); ind++) {
+        stars.push_back(options["LOCAL_IBO_STARS"][ind].to_integer() - 1);
+    }
+    local->set_stars(stars);
```
Why is this not in `common_init`?

The current content of `common_init` is irrelevant when you do this setting right after.

andyj10224 replied:
I am following the convention of the base Localizer class, which does the same thing (initializes instance variables in `common_init`, and then sets them with options in the `Localizer::build` method.

JonathonMisiewicz replied:
For future reference, "following the convention" is often not a good argument if there's not a good reason for the convention.

But in this case, there is a reason that I missed the first time. Add a comment in the header file saying that the user will normally set options by passing in an `options` object through `build`, but options are initialized in `common_init` in case that doesn't happen.

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+
+    // => Overlap Integrals <= //
+
+    auto fact11 = std::make_shared<IntegralFactory>(primary_, primary_, primary_, primary_);
+    auto fact12 = std::make_shared<IntegralFactory>(primary_, minao_, primary_, minao_);
+    auto fact22 = std::make_shared<IntegralFactory>(minao_, minao_, minao_, minao_);
+
+    std::shared_ptr<OneBodyAOInt> ints11(fact11->ao_overlap());
+    std::shared_ptr<OneBodyAOInt> ints12(fact12->ao_overlap());
+    std::shared_ptr<OneBodyAOInt> ints22(fact22->ao_overlap());
```
Please change to `unique`. These could be raw objects if we didn't reset.

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+
+    // => Tilde C <= //
+
+    std::shared_ptr<Matrix> C = C_;
```
There's a ton of `auto` to use from here on.

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+    for (int iter = 1; iter <= maxiter; iter++) {
+        double metric = 0.0;
+        for (int i = 0; i < nocc; i++) {
+            for (int A = 0; A < minao_inds.size(); A++) {
```
Use range-based for loops.

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+
+            double Aij = 0.0;
+            double Bij = 0.0;
+            for (int A = 0; A < minao_inds.size(); A++) {
```
Range-based for loops.

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+    int nocc = L->rowspi()[0];
+    int natom = true_atoms_.size();
+
+    std::shared_ptr<Matrix> Q = orbital_charges(L);
```
There is no reason for `Q`, `N`, or `A` to be a pointer. Change these to be raw objects. That'll mean changing the signature of `orbital_charges`.

andyj10224 replied:
I think Q (orbital_charges) should remain a `shared_ptr`, since the fsapt code needs to access the variable, and we don't wanna copy matrices.

JonathonMisiewicz replied:
I don't see how a copy would occur.

For expressions like `variable_type my_variable = my_function(my_arguments)`, C++ is smart enough not to create the return value and then copy it to the variable. It's called _return value optimization_.

andyj10224 replied:
I changed N to be a raw object, but not Q, since the fisapt code expects Q to be a pointer, and not a raw object.

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+        }
+        outfile->Printf("\n\n");
+
+        std::map<std::string, std::shared_ptr<Matrix> > ret2 =
```
`auto`

JonathonMisiewicz reviewed (commented):
This is mostly moving a class over, so I don't have many comments, but there is a design decision that makes no sense to me post-move, and quite a lot of type nitpicking.

IMO, combining Pipek-Mizey and IBO localization is a refactor outside the scope of this PR.

zachglick commented on psi4/src/psi4/libmints/local.h:
```diff
@@ -154,6 +155,108 @@ class PSI_API PMLocalizer : public Localizer {
+    // => Computers <= //
+
+    /// Print out the localization algorithm and parameters
+    virtual void print_header() const;
```
Same as the above destructor

zachglick reviewed (commented):
Thanks for taking this on Andy! It will be very useful to have IBOs outside of the FSAPT code.

I imagine a lot of psi4 users will want to export and visualize local orbitals. For this reason, we definitely want the IBOLocalizer exposed py-side as well, same as the other two localizers. See `export_mints.cc`

Something else we should think about (other devs please chime in) is how the various derived `Localizer` classes can be modified to have as similar method signatures as possible. The `BoysLocalizer` and `PMLocalizer` have a `void`-type `localize()` function and accessors `L()` and `U()`. The `IBOLocalizer` class has no accessors and instead returns a map of info from the `localize` function. I'm not sure which of these designs I prefer, but I think the three derived classes should at least be consistent.

Lastly, I noticed that there aren't really any tests or docs on the various localizers (except for class autodocs). It would be super nice if we could get one or both of those.

hokru commented:
> The BoysLocalizer and PMLocalizer have a void-type localize() function and accessors L() and U(). The IBOLocalizer class has no accessors and instead returns a map of info from the localize function. I'm not sure which of these designs I prefer, but I think the three derived classes should at least be consistent.

Agree. Consistency is crucial and within the PR scope. I think the accessors design is more common in psi4 but other than that I have no strong opinion on either choice.

JonathonMisiewicz commented on psi4/src/psi4/libmints/local.h:
```diff
@@ -154,6 +155,108 @@ class PSI_API PMLocalizer : public Localizer {
+
+    IBOLocalizer(std::shared_ptr<BasisSet> primary, std::shared_ptr<BasisSet> minao, std::shared_ptr<Matrix> C);
+
+    virtual ~IBOLocalizer();
```
For that matter, I don't think any of the subclass destructors are necessary. They're identical to the base class destructor.

susilehtola replied:
@zachglick the virtual is important if you make a child class of this class; otherwise the child class can't override it anymore. Right?

zachglick replied:
That sounds right @susilehtola. I don't think we anticipate making a child class of `IBOLocalizer`, so we should be okay

JonathonMisiewicz commented:
All good points from Zach.

I slightly prefer map return style; it makes it more transparent that the `Localizer` class is a glorified wrapper on the `localize` function. But my opinions are loosely held.

Incidentally, `localize_task` should be `localize` for sake of consistency.

andyj10224 commented:
> Thanks for taking this on Andy! It will be very useful to have IBOs outside of the FSAPT code.
> 
> I imagine a lot of psi4 users will want to export and visualize local orbitals. For this reason, we definitely want the IBOLocalizer exposed py-side as well, same as the other two localizers. See `export_mints.cc`
> 
> Something else we should think about (other devs please chime in) is how the various derived `Localizer` classes can be modified to have as similar method signatures as possible. The `BoysLocalizer` and `PMLocalizer` have a `void`-type `localize()` function and accessors `L()` and `U()`. The `IBOLocalizer` class has no accessors and instead returns a map of info from the `localize` function. I'm not sure which of these designs I prefer, but I think the three derived classes should at least be consistent.
> 
> Lastly, I noticed that there aren't really any tests or docs on the various localizers (except for class autodocs). It would be super nice if we could get one or both of those.

Hey Zach and all, I just updated the code to conform to the convention of the other localizers. I split the old multi-argument `localize` function in `IBOLocalizer` into the `localize` and `update_fock` methods inherited from the Localizer base class.

JonathonMisiewicz commented:
I'll leave a first pass on this to Zach, but a couple notes:
* Do _not_ mark a comment as resolved unless you've actually resolved it. I see a few cases where you "resolved" some of my comments but didn't make any code changes or leave comments as to why not.
* You'll need to fix the merge conflicts with FSAPT. It's probably just the change in copyright date.

zachglick commented on psi4/src/psi4/fisapt/fisapt.cc:
```diff
@@ -118,13 +118,13 @@ void FISAPT::localize() {
-        fisapt::IBOLocalizer2::build(primary_, reference_->get_basisset("MINAO"), matrices_["Cocc"], options_);
-    local->print_header();
-    std::map<std::string, std::shared_ptr<Matrix> > ret = local->localize(matrices_["Cocc"], Focc, ranges);
+    auto local = IBOLocalizer::build(primary_, reference_->get_basisset("MINAO"), matrices_["Cocc"], options_);
+    local->set_ranges(ranges);
+    local->localize();
+    local->fock_update(Focc);
```
It appears that the `IBOLocalizer::build` call is always accompanied by exactly one call to `set_ranges` before localization and exactly one call to `fock_update` after localization. For this reason, I don't think `set_ranges` or `fock_update` should be user facing functions. They don't add any flexibility to the user, and they require the user to memorize a list of additional functional calls.

Perhaps `ranges` and `Focc` can be made arguments to `build`, `set_ranges` can be made to call `build`, and `localize` can be made to also call `fock_update`.

andyj10224 replied:
Addressed

zachglick commented on psi4/src/psi4/libmints/local.h:
```diff
@@ -154,6 +155,106 @@ class PSI_API PMLocalizer : public Localizer {
+
+    /// Localize function from parent class
+    void localize() override;
+    /// Fock update function for IBO (Slightly different than for regular localizers)
```
An informative comment would be helpful here

zachglick reviewed (commented):
This is looking better Andy! I left a couple of comments. I think you might've missed part of my original review. We need to be able to export these orbitals py-side and we'll also need at least one test to demonstrate that the exporting works okay.

andyj10224 commented on psi4/src/psi4/libmints/local.cc:
```diff
@@ -549,4 +550,527 @@ void PMLocalizer::localize() {
+    auto DSDtilde = linalg::triplet(D, S11, Dtilde, false, false, false);
+    DSDtilde->scale(2.0);
+
+    auto L = linalg::doublet(S11_m12, S11_m12, false, false);  // TODO: Possibly Unstable
```
This was pre-existing

JonathonMisiewicz commented:
Ping me when this could use another review from me. From tests, it looks like there's a bug in the current code.
